core/src/main/scala/kafka/server/AddPartitionsToTxnManager.scala

```diff
@@ -0,0 +1,170 @@
+            }
+          }
+          val oldCallback = currentNodeAndTransactionData.callbacks(transactionData.transactionalId())
+          currentNodeAndTransactionData.transactionData.remove(transactionData)
+          oldCallback(topicPartitionsToError.toMap)
+        } else {
+          // We should never see a request on the same epoch since we haven't finished handling the one in queue
+          throw new InvalidRecordException("Received a second request from the same connection without finishing the first.")
+        }
+      }
+      currentNodeAndTransactionData.transactionData.add(transactionData)
+      currentNodeAndTransactionData.callbacks.put(transactionData.transactionalId(), callback)
+      wakeup()
+    }
+  }
+
+  private class AddPartitionsToTxnHandler(node: Node, transactionDataAndCallbacks: TransactionDataAndCallbacks) extends RequestCompletionHandler {
+    override def onComplete(response: ClientResponse): Unit = {
+      inflightNodes.remove(node)
+      if (response.authenticationException() != null) {
+        error(s"AddPartitionsToTxnRequest failed for broker ${config.brokerId} with an " +
+          "authentication exception.", response.authenticationException)
+        transactionDataAndCallbacks.callbacks.foreach { case (txnId, callback) =>
+          callback(buildErrorMap(txnId, transactionDataAndCallbacks.transactionData, Errors.forException(response.authenticationException()).code()))
+        }
+      } else if (response.versionMismatch != null) {
+        // We may see unsupported version exception if we try to send a verify only request to a broker that can't handle it. 
+        // In this case, skip verification.
+        error(s"AddPartitionsToTxnRequest failed for broker ${config.brokerId} with invalid version exception. This suggests verification is not supported." +
+              s"Continuing handling the produce request.")
+        transactionDataAndCallbacks.callbacks.values.foreach(_(Map.empty))
+      } else {
+        val addPartitionsToTxnResponseData = response.responseBody.asInstanceOf[AddPartitionsToTxnResponse].data
+        if (addPartitionsToTxnResponseData.errorCode != 0) {
+          error(s"AddPartitionsToTxnRequest for broker ${config.brokerId}  returned with error ${Errors.forCode(addPartitionsToTxnResponseData.errorCode)}.")
+          // TODO: send error back correctly -- we need to verify all possible errors can be handled by the client.
+          // errors -- versionmismatch --> handled above
+          //        -- clusterauth --> should handle differently
+          transactionDataAndCallbacks.callbacks.foreach { case (txnId, callback) =>
+            callback(buildErrorMap(txnId, transactionDataAndCallbacks.transactionData, addPartitionsToTxnResponseData.errorCode()))
+          }
+        } else {
+          addPartitionsToTxnResponseData.resultsByTransaction().forEach { transactionResult =>
```

Comment:
The KIP also mentions a race condition where the transaction may be aborted just after we've verified it, but before we got the reply, it doesn't seem like this PR addresses that, do we plan to do it in a different PR?

Reply:
thanks for the reminder.

Reply:
I've changed the internal details a bit so I no longer have an identifier that the first check succeeded. We append to the log on callback. Originally, I was going to update the PSM entry to contain the first offset and then we would remove the entry on the end marker. However, now we don't write the offset until the first append. (We can check for the next ones, but not the first append). The later changes with the epochs will make this not necessary since we bump the epoch on the marker write. I'm trying to think how we can handle this in the old clients case on the first append -- if there are any indicators we can check.

Reply:
One option here is to check the timestamp of the last time the entry was updated if there is no "last txn" information. On the first append, the timestamp should not change on the entry. (I believe it only changes on data appends and txn markers)

Reply:
If we don't want to put too many things in one change, we could implement the race condition checks in a separate change -- even though we didn't fully fix the problem we didn't regress (in fact improved quite a bit).  On the other hand, fixing localTime metric should be done in this change, because it worked before this change so if we don't fix it, it would be a regression.

Another approach (if it makes things simpler in any way) could be to split out the framework to run callbacks on request threads and add metrics there, then rebase this change on top of framework, so this change focuses on the transaction-specific stuff.

Reply:
Let's do the check before append in a followup -- here is the JIRA -- https://issues.apache.org/jira/browse/KAFKA-14561